src/main/java/com/google/starfish/servlets/HandleNotesServlet.java

```diff
@@ -1,58 +1,128 @@
 @WebServlet("/handle-notes")
 public class HandleNotesServlet extends HttpServlet {
 
   private BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
   private final String COOKIE_NAME = "SFCookie";
+  private LabelService labelService = new LabelService();
+  private MiscNoteLabelService miscNoteLabelService = new MiscNoteLabelService();
   
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
     if(!validateUser(request)) {
       response.setStatus(HttpServletResponse.SC_FORBIDDEN);
       return;
     }
 
     HttpSession activeSession = request.getSession(false);
     String userId = (String) activeSession.getAttribute("user_id");
-
     String blobKey = getUploadedFileBlobKey(request, "file");
-    if(blobKey == null) {
+    String sourceUrl = request.getParameter("sourceUrl");
+
+    // There must be some source, either a blob or url, for this note to be posted
+    if (blobKey == null && sourceUrl == null) {
       response.setStatus(HttpServletResponse.SC_NOT_ACCEPTABLE);
       return;
     }
-    String title = request.getParameter("title").toLowerCase();
+
+    String title = request.getParameter("title");
     String school = request.getParameter("school").toLowerCase();
     String course = request.getParameter("course").toLowerCase();
-    String[] miscLabels = request.getParameter("miscLabels").split(",");
-
-    // TODO: Put labels into database
+    String[] miscLabels = request.getParameterValues("miscLabels");
 
-    // TODO: Put note into database
+    DataSource pool = (DataSource) request.getServletContext().getAttribute("my-pool");
 
-    response.setStatus(HttpServletResponse.SC_OK);
+    // Insert all labels before inserting note due to foreign key constraint
+    labelService.insertSchoolLabel(pool, school);
+    labelService.insertCourseLabel(pool, course);
+    for (String miscLabel : miscLabels) {
+      labelService.insertMiscLabel(pool, miscLabel.toLowerCase());
+    }
+    
+    try (Connection conn = pool.getConnection()) {
+      String stmt =
+          "INSERT INTO notes ( "
+              + "author_id,"
+              + "school,"
+              + "course,"
+              + "title,"
+              + "source_url,"
+              + "pdf_source,"
+              + "date_created,"
+              + "num_downloads ) "
+        + "VALUES ( "
+              + "?,"
+              + "?,"
+              + "?,"
+              + "?,"
+              + "?,"
+              + "?,"
+              + "?,"
+              + "? ); ";
+
+      try (PreparedStatement noteStmt = conn.prepareStatement(stmt, new String[] {"id"})) {
+        noteStmt.setString(1, userId);
+        noteStmt.setString(2, school);
+        noteStmt.setString(3, course);
+        noteStmt.setString(4, title);
+        if (sourceUrl == null) {
```

Comment:
It feels odd to check for sourceUrl as the requirement to add a blobkey. Could we instead check for a sourceUrl to set the source_url as sourceUrl?

Reply:
Yep, I'll check for `sourceUrl != null` instead!